### PR TITLE
ActivitySource minor optimization and test

### LINKED ISSUE #1075: ActivitySourceAdapter fixes/improvements
1. `ActivitySourceAdapter` is [taking a snapshot of the processor](https://github.com/open-telemetry/opentelemetry-dotnet/blob/812fe936cf70e7ac72769e1e4074d54babea4dea/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L58), which is a bug since the processor can be [altered later](https://github.com/open-telemetry/opentelemetry-dotnet/blob/812fe936cf70e7ac72769e1e4074d54babea4dea/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L153).
2. `ActivitySourceAdapter` [doesn't have special case perf optimization](https://github.com/open-telemetry/opentelemetry-dotnet/blob/812fe936cf70e7ac72769e1e4074d54babea4dea/src/OpenTelemetry/Trace/ActivitySourceAdapter.cs#L74) for built-in samplers  Probably worth to do the same optimization as #1069.

### PR DESCRIPTION
**Pull request by cijothomas**
Fixes #1075 Part 2.

Also added a null check to ActivitySource's processor. 
The testing require some refactor to properly test individual component independently. WIll do a future PR to refactor unit testing.

For significant contributions please make sure you have completed the following items:

* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/Trace/ActivitySourceAdapter.cs [hunk 1]
```diff
@@ -84,47 +84,59 @@ internal void UpdateProcessor(ActivityProcessor processor)
...
-            ActivityContext parentContext;
-            if (string.IsNullOrEmpty(activity.ParentId))
+            if (this.sampler is AlwaysOnSampler)
             {
-                parentContext = default;
+                activity.IsAllDataRequested = true;
+                activity.ActivityTraceFlags |= ActivityTraceFlags.Recorded;
             }
-            else if (activity.Parent != null)
+            else if (this.sampler is AlwaysOffSampler)
```
Comment (reyang) [inline src/OpenTelemetry/Trace/ActivitySourceAdapter.cs:92]: Is it possible to hardwire this piece so we only check `this.sampler is AlwaysOnSampler` once and save all subsequent checks entirely?

Comment (reyang) [reply src/OpenTelemetry/Trace/ActivitySourceAdapter.cs:92]: @cijothomas probably check with @rajkumar-rangaraj.

Comment (cijothomas) [reply src/OpenTelemetry/Trace/ActivitySourceAdapter.cs]: @reyang Made some improvements. Let me know if this looks good. I will createbenchmark and try out later.

File: src/OpenTelemetry/Trace/ActivitySourceAdapter.cs [hunk 2]
```diff
@@ -39,10 +40,34 @@ public class ActivitySourceAdapter
...
+
+            if (resource == null)
+            {
+                throw new ArgumentNullException(nameof(resource));
+            }
+
             this.sampler = sampler;
+            if (this.sampler is AlwaysOnSampler)
+            {
+                this.getRequestedDataAction = this.RunGetRequestedDataAlwaysOnSampler;
```
Comment (reyang) [inline src/OpenTelemetry/Trace/ActivitySourceAdapter.cs:60]: probably a lambda here?

Comment (cijothomas) [reply src/OpenTelemetry/Trace/ActivitySourceAdapter.cs:60]: Will do after my benchmark

**Pull request merged**